examples/httpassembly/main_bpf.go

```diff
@@ -0,0 +1,214 @@
+
+// httpStreamFactory implements tcpassembly.StreamFactory
+type httpStreamFactory struct{}
+
+// httpStream will handle the actual decoding of http requests.
+type httpStream struct {
+	net, transport gopacket.Flow
+	r              tcpreader.ReaderStream
+}
+
+func (h *httpStreamFactory) New(net, transport gopacket.Flow) tcpassembly.Stream {
+	hstream := &httpStream{
+		net:       net,
+		transport: transport,
+		r:         tcpreader.NewReaderStream(),
+	}
+	go hstream.run() // Important... we must guarantee that data from the reader stream is read.
+
+	// ReaderStream implements tcpassembly.Stream, so we can return a pointer to it.
+	return &hstream.r
+}
+
+func (h *httpStream) run() {
+	buf := bufio.NewReader(&h.r)
+	for {
+		req, err := http.ReadRequest(buf)
+		if err == io.EOF {
+			// We must read until we see an EOF... very important!
+			return
+		} else if err != nil {
+			log.Println("Error reading stream", h.net, h.transport, ":", err)
+		} else {
+			bodyBytes := tcpreader.DiscardBytesToEOF(req.Body)
+			req.Body.Close()
+			//log.Println("Received request from stream", h.net, h.transport, ":", req, "with", bodyBytes, "bytes in request body")
+			log.Println(h.net, h.transport, req.Method, req.Proto, req.Host, req.URL, bodyBytes)
+		}
+	}
+}
+
+func main() {
+	defer util.Run()()
+	//var handle *pcap.Handle
+	var handle *bsdbpf.BPFSniffer
+	var err error
+	var options = bsdbpf.Options{
+		BPFDeviceName: "",
+		//ReadBufLen:       32767,
+		ReadBufLen: 0, // asks BPF for buffer size (32768 with OpenBSD 5.7)
+		//Timeout:          nil,
+		Timeout: &syscall.Timeval{Sec: 1, Usec: 0},
+		Promisc: true,
+		//Immediate:      true,
+		Immediate:        false,
+		PreserveLinkAddr: true,
+	}
+
+	// Set up pcap packet capture
+	if *fname != "" {
+		log.Printf("Reading from pcap dump %q", *fname)
+		//handle, err = pcap.OpenOffline(*fname)
+		log.Fatal("Reading from pcap dump %q not yet implemented on BSD", *fname)
+	} else {
+		log.Printf("Starting capture on interface %q", *iface)
+		//handle, err = pcap.OpenLive(*iface, int32(*snaplen), true, pcap.BlockForever)
+		//handle, err = bsdbpf.NewBPFSniffer(*iface, nil)
+		handle, err = bsdbpf.NewBPFSniffer(*iface, &options)
+	}
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	if err := handle.SetBpfReadFilterProgram(bpfHTTPFilterProg); err != nil {
+		log.Fatal(err)
+	}
+	if err := handle.FlushBpf(); err != nil {
+		log.Fatal(err)
+	}
+
+	// Set up assembly
+	streamFactory := &httpStreamFactory{}
+	streamPool := tcpassembly.NewStreamPool(streamFactory)
+	assembler := tcpassembly.NewAssembler(streamPool)
+
+	log.Println("reading in packets")
+	// Read in packets, pass to assembler.
+	//packetSource := gopacket.NewPacketSource(handle, handle.LinkType())
+	packetSource := gopacket.NewPacketSource(handle, layers.LayerTypeEthernet)
+	packets := packetSource.Packets()
+	ticker := time.Tick(time.Minute)
+
+	for {
```

Comment:
Can we factor a bunch of this logic out into a third file, then call process(packetSoruce) once we have the packet source?